src/Cardano/Shell/DaedalusIPC.hs

```diff
@@ -0,0 +1,233 @@
+                                  runPut)
+import           Data.Functor (($>))
+import           Data.Maybe (fromMaybe)
+import           Data.Text (Text)
+import           Data.Word (Word32, Word64)
+
+import           GHC.IO.Handle.FD (fdToHandle)
+import           System.Environment (lookupEnv)
+import           System.Info (arch)
+import           System.IO (Handle, hFlush, hGetLine, hSetNewlineMode,
+                            noNewlineTranslation)
+import           System.IO.Error (IOError, userError)
+import           Text.Read (readEither)
+
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.ByteString.Lazy.Char8 as L8
+import qualified Data.Text as T
+
+----------------------------------------------------------------------------
+-- Daedalus <-> Wallet child process port discovery protocol
+
+data MsgIn = QueryPort
+    deriving (Show, Eq)
+data MsgOut = Started | ReplyPort Int | ParseError Text
+    deriving (Show, Eq)
+
+instance FromJSON MsgIn where
+    parseJSON = withObject "MsgIn" $ \v -> do
+        (_ :: [()]) <- v .: "QueryPort"
+        pure QueryPort
+
+instance ToJSON MsgOut where
+    toJSON Started        = object [ "Started" .= Array mempty ]
+    toJSON (ReplyPort p)  = object [ "ReplyPort" .= p ]
+    toJSON (ParseError e) = object [ "ParseError" .= e ]
+
+-- | Start up the Daedalus IPC process. It's called 'daedalusIPC', but this
+-- could be any nodejs program that needs to start cardano-wallet. All it does
+-- is reply with a port number when asked, using a very nodejs-specific IPC
+-- method.
+--
+-- If the IPC channel was successfully set up, this function won't return until
+-- the parent process exits. Otherwise, it will return immediately. Before
+-- returning, it will log an message about why it has exited.
+daedalusIPC
+    :: Trace IO Text
+    -- ^ Logging object
+    -> Int
+    -- ^ Port number to send to Daedalus
+    -> IO ()
+daedalusIPC trace port = withNodeChannel (pure . msg) action >>= \case
+    Right runServer -> do
+        logInfo trace "Daedalus IPC server starting"
+        runServer >>= \case
+            Left (NodeChannelFinished err) ->
+                logNotice trace $ "Daedalus IPC finished for this reason: " <> show err
+            Right () -> logError trace "Unreachable code"
+    Left NodeChannelDisabled -> do
+        logInfo trace "Daedalus IPC is not enabled."
+        sleep
+    Left (NodeChannelBadFD err) ->
+        logError trace $ "Problem starting Daedalus IPC: " <> show err
+  where
+    -- How to respond to an incoming message, or when there is an incoming
+    -- message that couldn't be parsed.
+    msg (Right QueryPort) = Just (ReplyPort port)
+    msg (Left e)          = Just (ParseError e)
+
+    -- What to do in context of withNodeChannel
+    action :: (MsgOut -> IO ()) -> IO ()
+    action send = send Started >> sleep
+
+    sleep = threadDelay maxBound
+
+----------------------------------------------------------------------------
```

Comment:
Everything below this line is not specific to Daedalus and is just for the nodejs ipc protocol. It could go in a separate module.